README.md
# ExtDate

Native JavaScript Date object extended with new useful methods.

[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square)](http://standardjs.com/)
[![latest-release](https://img.shields.io/github/release/MrSlide/ExtDate.svg?style=flat-square)](https://github.com/MrSlide/ExtDate/tree/master)
[![Travis branch](https://img.shields.io/travis/MrSlide/ExtDate/master.svg?style=flat-square)](https://travis-ci.org/MrSlide/ExtDate)
[![Coverage Status](https://img.shields.io/coveralls/MrSlide/ExtDate/master.svg?style=flat-square)](https://coveralls.io/github/MrSlide/ExtDate?branch=master)
[![GitHub issues](https://img.shields.io/github/issues/MrSlide/ExtDate.svg?style=flat-square)](https://github.com/MrSlide/ExtDate/issues)
[![license](https://img.shields.io/github/license/MrSlide/ExtDate.svg?style=flat-square)](https://opensource.org/licenses/MIT)



- [Installation](#installation)
  - [Via Bower](#via-bower)
  - [Via NPM](#via-npm)
- [Usage](#usage)
  - [Via ES6 syntax](#via-es6-syntax)
  - [Via CommonJS syntax](#via-commonjs-syntax)
  - [Via the script tag](#via-the-script-tag)
- [API](#api)
  - [ExtDate()](#extdate-1)
  - [ExtDate.prototype.constructor()](#extdateprototypeconstructor)
  - [ExtDate.prototype.getDayOfYear()](#extdateprototypegetdayofyear)
  - [ExtDate.prototype.getUTCDayOfYear()](#extdateprototypegetutcdayofyear)
  - [ExtDate.prototype.getWeek()](#extdateprototypegetweek)
  - [ExtDate.prototype.getUTCWeek()](#extdateprototypegetutcweek)
  - [ExtDate.prototype.setDayOfYear()](#extdateprototypesetdayofyearnumber-day-number-year)
  - [ExtDate.prototype.setUTCDayOfYear()](#extdateprototypesetutcdayofyearnumber-day-number-year)
  - [ExtDate.prototype.setFirstDayOfYear()](#extdateprototypesetfirstdayofyearnumber-year)
  - [ExtDate.prototype.setUTCFirstDayOfYear()](#extdateprototypesetutcfirstdayofyearnumber-year)
  - [ExtDate.prototype.setFirstWeekOfYear()](#extdateprototypesetfirstweekofyearnumber-year)
  - [ExtDate.prototype.setUTCFirstWeekOfYear()](#extdateprototypesetutcfirstweekofyearnumber-year)
  - [ExtDate.prototype.setStartOfDay()](#extdateprototypesetstartofday)
  - [ExtDate.prototype.setUTCStartOfDay()](#extdateprototypesetutcstartofday)
  - [ExtDate.prototype.setWeek()](#extdateprototypesetweeknumber-week-number-year)
  - [ExtDate.prototype.setUTCWeek()](#extdateprototypesetutcweeknumber-week-number-year)
- [Brower support](#browser-support)
- [Support](#support)
- [Authors and Contributors](#authors-and-contributers)
- [License and copyright](#license-and-copyright)



## Installation

### Via [Bower](http://bower.io/)

```
bower install ext-date
```

### Via [NPM](https://www.npmjs.com/)

```
npm install ext-date
```


## Usage

ExtDate is an [UMD](https://github.com/umdjs/umd) module. You can load it into your application either by importing the module, or loading the script in your page.

If you are importing the ExtDate module via [Webpack](https://webpack.github.io/), [Browserify](http://browserify.org/) or similar, make sure that the module name `ext-date` is being resolved correctly to the [Bower](http://bower.io/) or [NPM](https://www.npmjs.com/) packages folder.

### Via ES6 syntax

```
import ExtDate from 'ext-date'
```

### Via CommonJS syntax

```
var ExtDate = require('ext-date')
```

### Via the script tag

```
<script src="/scripts/ExtDate.js"></script> // Change the path as necessary
```



## API

The ExtDate API tries to mimic the native [Date] object as much as possible. All the static methods and instance methods available in the native [Date] object will be available through ExtDate, as long as they are supported by the browser.

For documentation on the native methods, please see <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date>.

For a consistent behaviour, if ExtDate has any method that also exists in the native [Date] object, the ExtDate method will take precedence. ExtDate does not modify the native [Date] object, so you can still access the any native method by using the native [Date] directly instead of ExtDate.



### ExtDate()

Returns a string representation of the current date and time.

**Parameters**

See <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date>.

**Returns**

[String] A string representation of the current date and time. Equivalent to `(new Date().toString())`.



### ExtDate.prototype.constructor()

Creates an instance of a [Date] object extended with the ExtDate methods.
The ExtDate constructor uses the same syntax as the native [Date] object constructor. See <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date> for more details and parameter options.


```
var dateInstance = new ExtDate()
var dateInstance = new ExtDate('December 17, 1995 03:24:00')
var dateInstance = new ExtDate('1995-12-17T03:24:00')
var dateInstance = new ExtDate(1995, 11, 17)
var dateInstance = new ExtDate(1995, 11, 17, 3, 24, 0)
```

**Parameters**

See <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date>.

**Returns**

[Date] A new Date instance extended with the ExtDate methods.



### ExtDate.prototype.getDayOfYear()

Returns the day of the year for a specified date according to local time.


```
var dayfOfYear = dateInstance.getDayOfYear()
```

**Returns**

[Number] The day of the year for the specified date.



### ExtDate.prototype.getUTCDayOfYear()

Returns the day of the year for a specified date according to universal time.


```
var dayfOfYear = dateInstance.getUTCDayOfYear()
```

**Returns**

[Number] The day of the year for the specified date.



### ExtDate.prototype.getWeek()

Gets the week number of the specified date according to local time.


```
var weekNumber = dateInstance.getWeek()
```

**Returns**

[Number] The ISO week number.



### ExtDate.prototype.getUTCWeek()

Gets the week number of the specified date according to universal time.


```
var weekNumber = dateInstance.getUTCWeek()
```

**Returns**

[Number] The ISO week number.



### ExtDate.prototype.setDayOfYear([Number] day[, [Number] year])

Sets the day of the year for a specified date according to local time.


```
dateInstance.setDayOfYear(45, 2016)
```

**Parameters**

| Name | Type     | Required | Default              | Description                                                                                         |
|------|----------|----------|----------------------|-----------------------|
| day  | [Number] | `true`   |                      | An integer between 1 and 365 (or 366, if the year is a leap year) representing the day of the year. |
| year | [Number] | `false`  | `this.getFullYear()` | If not specified, the year of the specified date will remain unchanged.                             |

**Returns**

[Number] The number of milliseconds since 1 January 1970 00:00:00 UTC.



### ExtDate.prototype.setUTCDayOfYear([Number] day[, [Number] year])

Sets the day of the year for a specified date according to universal time.


```
dateInstance.setUTCDayOfYear(45, 2016)
```

**Parameters**

| Name | Type     | Required | Default                 | Description                                                                                         |
|------|----------|----------|-------------------------|-----------------------|
| day  | [Number] | `true`   |                         | An integer between 1 and 365 (or 366, if the year is a leap year) representing the day of the year. |
| year | [Number] | `false`  | `this.getUTCFullYear()` | If not specified, the year of the specified date will remain unchanged.                             |

**Returns**

[Number] The number of milliseconds since 1 January 1970 00:00:00 UTC.



### ExtDate.prototype.setFirstDayOfYear([[Number] year])

Sets the first day of the year for a specified date according to local time.


```
dateInstance.setFirstDayOfYear(2016)
```

**Parameters**

| Name | Type     | Required | Default              | Description                                                             |
|------|----------|----------|----------------------|----------------------------------|
| year | [Number] | `false`  | `this.getFullYear()` | If not specified, the year of the specified date will remain unchanged. |

**Returns**

[Number] The number of milliseconds since 1 January 1970 00:00:00 UTC.



### ExtDate.prototype.setUTCFirstDayOfYear([[Number] year])

Sets the first day of the year for a specified date according to universal time.


```
dateInstance.setUTCFirstDayOfYear(2016)
```

**Parameters**

| Name | Type     | Required | Default                 | Description                                                             |
|------|----------|----------|-------------------------|----------------------------------|
| year | [Number] | `false`  | `this.getUTCFullYear()` | If not specified, the year of the specified date will remain unchanged. |

**Returns**

[Number] The number of milliseconds since 1 January 1970 00:00:00 UTC.



### ExtDate.prototype.setFirstWeekOfYear([[Number] year])

Sets the first day of the first week of the year for a specified date according to local time.


```
dateInstance.setFirstWeekOfYear(2016)
```

**Parameters**

| Name | Type     | Required | Default              | Description                                                             |
|------|----------|----------|----------------------|----------------------------------|
| year | [Number] | `false`  | `this.getFullYear()` | If not specified, the year of the specified date will remain unchanged. |

**Returns**

[Number] The number of milliseconds since 1 January 1970 00:00:00 UTC.



### ExtDate.prototype.setUTCFirstWeekOfYear([[Number] year])

Sets the first day of the first week of the year for a specified date according to universal time.


```
dateInstance.setUTCFirstWeekOfYear(2016)
```

**Parameters**

| Name | Type     | Required | Default                 | Description                                                             |
|------|----------|----------|-------------------------|----------------------------------|
| year | [Number] | `false`  | `this.getUTCFullYear()` | If not specified, the year of the specified date will remain unchanged. |

**Returns**

[Number] The number of milliseconds since 1 January 1970 00:00:00 UTC.



### ExtDate.prototype.setStartOfDay()

Sets the time of the day to 00:00:00.000 for a specified date according to local time.


```
dateInstance.setStartOfDay()
```

**Returns**

[Number] The number of milliseconds since 1 January 1970 00:00:00 UTC.



### ExtDate.prototype.setUTCStartOfDay()

Sets the time of the day to 00:00:00.000 for a specified date according to universal time.


```
dateInstance.setUTCStartOfDay()
```

**Returns**

[Number] The number of milliseconds since 1 January 1970 00:00:00 UTC.



### ExtDate.prototype.setWeek([Number] week[, [Number] year])

Sets the first day of a week number for a specified date according to local time.


```
dateInstance.setWeek(2, 2016)
```

**Parameters**

| Name | Type     | Required | Default              | Description                                                             |
|------|----------|----------|----------------------|----------------------------------|
| week | [Number] | `true`   |                      | An integer between 1 and 53, representing the week.                     |
| year | [Number] | `false`  | `this.getFullYear()` | If not specified, the year of the specified date will remain unchanged. |

**Returns**

[Number] The number of milliseconds since 1 January 1970 00:00:00 UTC.



### ExtDate.prototype.setUTCWeek([Number] week[, [Number] year])

Sets the first day of a week number for a specified date according to universal time.


```
dateInstance.setUTCWeek(2, 2016)
```

**Parameters**

| Name | Type     | Required | Default                 | Description                                                             |
|------|----------|----------|-------------------------|----------------------------------|
| week | [Number] | `true`   |                         | An integer between 1 and 53, representing the week.                     |
| year | [Number] | `false`  | `this.getUTCFullYear()` | If not specified, the year of the specified date will remain unchanged. |

**Returns**

[Number] The number of milliseconds since 1 January 1970 00:00:00 UTC.



## Brower support

ExtDate, although not tested in all of them, was created using features supported by these browsers.

- Android Browser 4+
- Blackberry Browser 7+
- Chrome 13+
- Firefox 4+
- Internet Explorer 9+
- Opera 12+
- Opera Mini 5+
- Safari 7+



## Support

If you want to request new features or find any bugs, please open a ticket on the [issues](https://github.com/MrSlide/ExtDate/issues) page and I'll review it as soon as possible.



## Authors and Contributors

Created by Luís Rodrigues ([@MrSlide](https://github.com/MrSlide))



## License and copyright

Released under the [MIT](https://opensource.org/licenses/MIT) license

Copyright (c) 2016 Luís Rodrigues


[Date]: https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Date
[String]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String
[Number]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [v1.0.0] - 2016-08-19

### Breaking changes

- Modified the bundled module to export the ExtDate class directly to the ExtDate global variable.


## [v0.1.0] - 2016-07-17

### Added

- Added getDayOfYear() and getUTCDayOfYear() methods to get the ordinal day of the year of a date instance.
- Added setDayOfYear() and setUTCDayOfYear() methods to set a date instance to an ordinal day of the year.
- Added getWeek() and getUTCWeek() methods to get the week number of a date instance.
- Added setWeek() and setUTCWeek() methods to set a date instance to a week number.
- Added setFirstDayOfYear() and setUTCFirstDayOfYear() methods to set a date instance to the start of the 1st of January.
- Added setFirstWeekOfYear() and setUTCFirstWeekOfYear() methods to set a date instance to the first week of a year.
- Added setStartOfDay() and setUTCStartOfDay() methods to set a date instance to the time 00:00:00.000.

[Latest]: https://github.com/MrSlide/ExtDate/tree/master
[Unreleased]: https://github.com/MrSlide/ExtDate/tree/develop
[v1.0.0]: https://github.com/MrSlide/ExtDate/tree/v1.0.0
[v0.1.0]: https://github.com/MrSlide/ExtDate/tree/v0.1.0


tests/unit-tests.js
const testsContext = require.context('../src', true, /\.spec\.js$/)
const moduleContext = require.context('../src', true, /^(.(?!\.spec))*\.js$/)

testsContext.keys().forEach(testsContext)
moduleContext.keys().forEach(moduleContext)


karma.conf.js
var path = require('path')
var configVars = require('./configVars.js')
var karmaConfig = {
  basePath: './',
  frameworks: ['jasmine-jquery', 'jasmine'],
  port: 9876,
  colors: true,
  autoWatch: false,
  browsers: ['PhantomJS'],
  singleRun: true,
  concurrency: 1,
  files: [
    './' + configVars['unitTests']['entryFile']
  ],
  reporters: [
    'spec',
    'coverage'
  ],
  specReporter: {
    maxLogLines: 2,
    suppressErrorSummary: false,
    suppressFailed: false,
    suppressPassed: false,
    suppressSkipped: true
  },
  coverageReporter: {
    includeAllSources: true,
    reporters: [
      {
        type: 'lcov',
        subdir: '../' + configVars['unitTests']['coverageOutputPath']
      },
      {
        type: 'text-summary'
      }
    ]
  },
  preprocessors: {},
  webpack: {
    devtool: 'inline-source-map',
    babel: {
      presets: ['es2015'],
      plugins: ['transform-object-assign'],
      cacheDirectory: true
    },
    isparta: {
      embedSource: true,
      noAutoWrap: true,
      babel: {
        presets: ['es2015'],
        plugins: [
          'transform-object-assign',
          'transform-es2015-modules-umd'
        ]
      }
    },
    module: {
      preLoaders: [
        {
          test: /\.js$/,
          loader: 'babel'
        },
        {
          test: /^(.(?!\.spec))*\.js$/,
          include: path.resolve('./' + configVars['js']['srcPath']),
          loader: 'isparta'
        }
      ]
    },
    externals: [
      {}
    ],
    cache: true
  }
}
karmaConfig.preprocessors['./' + configVars['unitTests']['entryFile']] = ['webpack', 'sourcemap']

module.exports = function (config) {
  karmaConfig.logLevel = config.LOG_ERROR
  config.set(karmaConfig)
}


gulp/compileJs.js
var gulp = require('gulp')
var rollup = require('rollup')
var babel = require('rollup-plugin-babel')
var configVars = require('../configVars.js')

var cache
/**
 * Transpile and minify source JavaScript code using Webpack and Babel
 */
gulp.task('compileJs', function (cb) {
  rollup.rollup({
    entry: './' + configVars['js']['srcPath'] + configVars['js']['entryFile'],
    cache: cache,
    plugins: [
      babel({
        presets: [
          ['es2015', {modules: false}]
        ],
        plugins: [
          'transform-object-assign',
          'external-helpers'
        ],
        exclude: 'node_modules/**',
        comments: false
      })
    ]
  }).then(function (bundle) {
    cache = bundle

    return bundle.write({
      dest: './' + configVars['global']['outputPath'] + configVars['js']['outputPath'] + configVars['js']['outputFile'],
      format: 'umd',
      moduleName: configVars['js']['libraryName'],
      sourceMap: true
    })
  }).then(function () {
    cb()
  }, function (err) {
    console.error(err)
    process.exit(1)
  })
})


src/ExtDate.js
'use strict'

/**
 * @name ExtDate
 * @desc Extends the native Date object with new useful methods.
 * @author Luis Rodrigues (http://www.luisrodriguesweb.com)
 * @version 1.0.0
 * @license MIT
 */

const extDateMethods = {
  /**
   * Sets the time of the day to 00:00:00.000 for a specified date according to local or univeral time.
   *
   * @memberof ExtDate
   * @private
   * @param {?Boolean} [utc] - Set to `true` to perform operations in univeral time.
   * @return {Number} - The number of milliseconds since 1 January 1970 00:00:00 UTC.
   * @since 0.1.0
   */
  _setStartOfDay: function _setStartOfDay (utc) {
    utc = utc ? 'UTC' : ''

    this[`set${utc}Milliseconds`](0)
    this[`set${utc}Seconds`](0)
    this[`set${utc}Minutes`](0)
    this[`set${utc}Hours`](0)

    return this.getTime()
  },

  /**
   * Sets the time of the day to 00:00:00.000 for a specified date according to local time.
   *
   * @memberof ExtDate
   * @public
   * @return {Number} - The number of milliseconds since 1 January 1970 00:00:00 UTC.
   * @since 0.1.0
   *
   * @example
   * dateInstance.setStartOfDay()
   */
  setStartOfDay: function setStartOfDay () {
    return this._setStartOfDay()
  },

  /**
   * Sets the time of the day to 00:00:00.000 for a specified date according to universal time.
   *
   * @memberof ExtDate
   * @public
   * @return {Number} - The number of milliseconds since 1 January 1970 00:00:00 UTC.
   * @since 0.1.0
   *
   * @example
   * dateInstance.setUTCStartOfDay()
   */
  setUTCStartOfDay: function setUTCStartOfDay () {
    return this._setStartOfDay(true)
  },

  /**
   * Sets the first day of the year for a specified date according to local or univeral time.
   *
   * @memberof ExtDate
   * @private
   * @param {?Number} [year] - If not specified, the year of the specified date will remain unchanged.
   * @param {?Boolean} [utc] - Set to `true` to perform operations in univeral time.
   * @return {Number} - The number of milliseconds since 1 January 1970 00:00:00 UTC.
   * @since 0.1.0
   */
  _setFirstDayOfYear: function _setFirstDayOfYear (year, utc) {
    utc = utc ? 'UTC' : ''

    this._setStartOfDay(!!utc)

    this[`set${utc}Date`](1)
    this[`set${utc}Month`](0)

    if (year) {
      this[`set${utc}FullYear`](year)
    }

    return this.getTime()
  },

  /**
   * Sets the first day of the year for a specified date according to local time.
   *
   * @memberof ExtDate
   * @public
   * @param {?Number} [year] - If not specified, the year of the specified date will remain unchanged.
   * @return {Number} - The number of milliseconds since 1 January 1970 00:00:00 UTC.
   * @since 0.1.0
   *
   * @example
   * dateInstance.setFirstDayOfYear(2016)
   */
  setFirstDayOfYear: function setFirstDayOfYear (year) {
    return this._setFirstDayOfYear(year)
  },

  /**
   * Sets the first day of the year for a specified date according to universal time.
   *
   * @memberof ExtDate
   * @public
   * @param {?Number} [year] - If not specified, the year of the specified date will remain unchanged.
   * @return {Number} - The number of milliseconds since 1 January 1970 00:00:00 UTC.
   * @since 0.1.0
   *
   * @example
   * dateInstance.setUTCFirstDayOfYear(2016)
   */
  setUTCFirstDayOfYear: function setUTCFirstDayOfYear (year) {
    return this._setFirstDayOfYear(year, true)
  },

  /**
   * Returns the day of the year for a specified date according to local time.
   *
   * @memberof ExtDate
   * @private
   * @param {?Boolean} [utc] - Set to `true` to perform operations in univeral time.
   * @return {Number} - The day of the year for the specified date.
   * @since 0.1.0
   */
  _getDayOfYear: function _getDayOfYear (utc) {
    utc = utc ? 'UTC' : ''

    const helperDate = new Date(this.getTime())

    let dayCount = helperDate[`get${utc}Date`]()
    let i = helperDate[`get${utc}Month`]()

    while (i > 0) {
      helperDate[`set${utc}Date`](0)
      dayCount += helperDate[`get${utc}Date`]()

      i--
    }

    return dayCount
  },

  /**
   * Returns the day of the year for a specified date according to local time.
   *
   * @memberof ExtDate
   * @public
   * @return {Number} - The day of the year for the specified date.
   * @since 0.1.0
   *
   * @example
   * var dayfOfYear = dateInstance.getDayOfYear()
   */
  getDayOfYear: function getDayOfYear () {
    return this._getDayOfYear()
  },

  /**
   * Returns the day of the year for a specified date according to universal time.
   *
   * @memberof ExtDate
   * @public
   * @return {Number} - The day of the year for the specified date.
   * @since 0.1.0
   *
   * @example
   * var dayfOfYear = dateInstance.getUTCDayOfYear()
   */
  getUTCDayOfYear: function getUTCDayOfYear () {
    return this._getDayOfYear(true)
  },

  /**
   * Sets the day of the year for a specified date according to local or universal time.
   *
   * @memberof ExtDate
   * @private
   * @param {Number} day - An integer between 1 and 365 (or 366, if the year is a leap year) representing the day of the year.
   * @param {?Number} [year] - If not specified, the year of the specified date will remain unchanged.
   * @param {?Boolean} [utc] - Set to `true` to perform operations in univeral time.
   * @return {Number} - The number of milliseconds since 1 January 1970 00:00:00 UTC.
   * @since 0.1.0
   */
  _setDayOfYear: function _setDayOfYear (day, year, utc) {
    utc = utc ? 'UTC' : ''

    if (year) {
      this[`set${utc}FullYear`](year)
    }

    this._setFirstDayOfYear(year, !!utc)
    this[`set${utc}Date`](day)

    return this.getTime()
  },

  /**
   * Sets the day of the year for a specified date according to local time.
   *
   * @memberof ExtDate
   * @public
   * @param {Number} day - An integer between 1 and 365 (or 366, if the year is a leap year) representing the day of the year.
   * @param {?Number} [year] - If not specified, the year of the specified date will remain unchanged.
   * @return {Number} - The number of milliseconds since 1 January 1970 00:00:00 UTC.
   * @since 0.1.0
   *
   * @example
   * dateInstance.setDayOfYear(45, 2016)
   */
  setDayOfYear: function setDayOfYear (day, year) {
    return this._setDayOfYear(day, year)
  },

  /**
   * Sets the day of the year for a specified date according to universal time.
   *
   * @memberof ExtDate
   * @public
   * @param {Number} day - An integer between 1 and 365 (or 366, if the year is a leap year) representing the day of the year.
   * @param {?Number} [year] - If not specified, the year of the specified date will remain unchanged.
   * @return {Number} - The number of milliseconds since 1 January 1970 00:00:00 UTC.
   * @since 0.1.0
   *
   * @example
   * dateInstance.setUTCDayOfYear(45, 2016)
   */
  setUTCDayOfYear: function setUTCDayOfYear (day, year) {
    return this._setDayOfYear(day, year, true)
  },

  /**
   * Sets the first day of the first week of the year for a specified date according to local or univeral time.
   *
   * @memberof ExtDate
   * @private
   * @param {?Number} [year] - If not specified, the year of the specified date will remain unchanged.
   * @param {?Boolean} [utc] - Set to `true` to perform operations in univeral time.
   * @return {Number} - The number of milliseconds since 1 January 1970 00:00:00 UTC.
   * @since 0.1.0
   */
  _setFirstWeekOfYear: function _setFirstWeekOfYear (year, utc) {
    utc = utc ? 'UTC' : ''

    this._setFirstDayOfYear(year, !!utc)
    let offset

    // Weeks start on Mondays
    let day = this[`get${utc}Day`]() - 1

    if (day < 0) {
      day = 6
    }

    // The first week of the year must include the 4th of January
    if (day > 3) {
      offset = 7 - day
    } else {
      offset = -day
    }

    // Adjust the date with the calculated offset
    this[`set${utc}Date`](this[`get${utc}Date`]() + offset)

    return this.getTime()
  },

  /**
   * Sets the first day of the first week of the year for a specified date according to local time.
   *
   * @memberof ExtDate
   * @public
   * @param {?Number} [year] - If not specified, the year of the specified date will remain unchanged.
   * @return {Number} - The number of milliseconds since 1 January 1970 00:00:00 UTC.
   * @since 0.1.0
   *
   * @example
   * dateInstance.setFirstWeekOfYear(2016)
   */
  setFirstWeekOfYear: function setFirstWeekOfYear (year) {
    return this._setFirstWeekOfYear(year)
  },

  /**
   * Sets the first day of the first week of the year for a specified date according to universal time.
   *
   * @memberof ExtDate
   * @public
   * @param {?Number} [year] - If not specified, the year of the specified date will remain unchanged.
   * @return {Number} - The number of milliseconds since 1 January 1970 00:00:00 UTC.
   * @since 0.1.0
   *
   * @example
   * dateInstance.setUTCFirstWeekOfYear(2016)
   */
  setUTCFirstWeekOfYear: function setUTCFirstWeekOfYear (year) {
    return this._setFirstWeekOfYear(year, true)
  },

  /**
   * Sets the first day of a week number for a specified date according to local or universal time.
   *
   * @memberof ExtDate
   * @private
   * @param {Number} week - An integer between 1 and 53, representing the week.
   * @param {?Number} [year] - If not specified, the year of the specified date will remain unchanged.
   * @param {?Boolean} [utc] - Set to `true` to perform operations in univeral time.
   * @return {Number} - The number of milliseconds since 1 January 1970 00:00:00 UTC.
   * @since 0.1.0
   */
  _setWeek: function _setWeek (week, year, utc) {
    utc = utc ? 'UTC' : ''

    const offset = (week - 1) * 7

    this._setFirstWeekOfYear(year, !!utc)
    this[`set${utc}Date`](this[`get${utc}Date`]() + offset)

    return this.getTime()
  },

  /**
   * Sets the first day of a week number for a specified date according to local time.
   *
   * @memberof ExtDate
   * @public
   * @param {Number} week - An integer between 1 and 53, representing the week.
   * @param {?Number} [year] - If not specified, the year of the specified date will remain unchanged.
   * @return {Number} - The number of milliseconds since 1 January 1970 00:00:00 UTC.
   * @since 0.1.0
   *
   * @example
   * dateInstance.setWeek(2, 2016)
   */
  setWeek: function setWeek (week, year) {
    return this._setWeek(week, year)
  },

  /**
   * Sets the first day of a week number for a specified date according to universal time.
   *
   * @memberof ExtDate
   * @public
   * @param {Number} week - An integer between 1 and 53, representing the week.
   * @param {?Number} [year] - If not specified, the year of the specified date will remain unchanged.
   * @return {Number} - The number of milliseconds since 1 January 1970 00:00:00 UTC.
   * @since 0.1.0
   *
   * @example
   * dateInstance.setUTCWeek(2, 2016)
   */
  setUTCWeek: function setUTCWeek (week, year) {
    return this._setWeek(week, year, true)
  },

  /**
   * Gets the week number of the specified date according to local or universal time.
   *
   * @memberof ExtDate
   * @private
   * @param {?Boolean} [utc] - Set to `true` to perform operations in univeral time.
   * @return {Number} - The ISO week number.
   * @since 0.1.0
   */
  _getWeek: function _getWeek (utc) {
    utc = utc ? 'UTC' : ''

    const day = this._getDayOfYear(!!utc)
    const weekDay = this[`get${utc}Day`]() || 7
    const week = Math.floor((day - weekDay + 10) / 7)

    return week !== 53 ? week : 1
  },

  /**
   * Gets the week number of the specified date according to local time.
   *
   * @memberof ExtDate
   * @public
   * @return {Number} - The ISO week number.
   * @since 0.1.0
   *
   * @example
   * var weekNumber = dateInstance.getWeek()
   */
  getWeek: function getWeek () {
    return this._getWeek()
  },

  /**
   * Gets the week number of the specified date according to universal time.
   *
   * @memberof ExtDate
   * @public
   * @return {Number} - The ISO week number.
   * @since 0.1.0
   *
   * @example
   * var weekNumber = dateInstance.getUTCWeek()
   */
  getUTCWeek: function getUTCWeek () {
    return this._getWeek(true)
  }
}

/**
 * Copy the properties of one object into another.
 *
 * @param {Object} src - The object to get the properties from.
 * @param {Object} dest - The object to copy the properties into.
 * @return {Object} - The extended object
 * @since 0.1.0
 */
function extend (src, dest) {
  const keys = Object.keys(src)

  keys.forEach(function (prop) {
    Object.defineProperty(dest, prop, {
      value: src[prop],
      enumerable: false,
      configurable: false,
      writable: false
    })
  })

  return dest
}

/**
 * Creates an instance of ExtDate which implements new methods for
 * the Date object.
 *
 * @function
 * @constructs
 * @public
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date
 * @return {Date|String} - The Date instance with the added methods, or a string representation of the date if called as a function.
 * @since 0.1.0
 *
 * @example
 * var currentDateTime = ExtDate()
 *
 * @example
 * new ExtDate()
 * new ExtDate(547979648008)
 * new ExtDate('1995-12-17T03:24:00')
 * new ExtDate(1995, 11, 17)
 * new ExtDate(1995, 11, 17, 3, 24, 0)
 */
export default function ExtDate (year, month, day, hour, minutes, seconds, milliseconds) {
  if (!(this instanceof ExtDate)) {
    return Date(...arguments)
  }

  // Create a Date instance to use internally
  const date = new Date(...arguments)

  return extend(extDateMethods, date)
}

// Extend ExtDate with Date's static methods
const staticProps = Object.getOwnPropertyNames(Date)

staticProps.forEach(function (prop) {
  if (typeof Date[prop] === 'function') {
    Object.defineProperty(ExtDate, prop, {
      value: Date[prop],
      enumerable: false,
      configurable: false,
      writable: false
    })
  }
})
